### DISCUSSION TITLE
Allow `RetryHandler` to stop without throwing

### DISCUSSION THREAD

**Discussion by mikaelkaron** [nodejs/undici#2996] [Ideas]
At the moment `onHeaders` takes `throwOnError` into account [like this](https://github.com/nodejs/undici/blob/v6.10.0/lib/api/api-request.js#L100-L113):

```javascript
if (this.throwOnError && statusCode >= 400) {
  this.runInAsyncScope(getResolveErrorBodyCallback, null,
    { callback, body, contentType, statusCode, statusMessage, headers }
  )
} else {
  this.runInAsyncScope(callback, null, null, {
    statusCode,
    headers,
    trailers: this.trailers,
    opaque,
    body,
    context
  })
}
```

However, `onError` _always_ [passes an error](https://github.com/nodejs/undici/blob/v6.10.0/lib/api/api-request.js#L141):

```javascript
this.runInAsyncScope(callback, null, err, { opaque })
```

Would it make sense to have `onError` work similar to `onHeaders` when calling `callback`:

```javascript
onError (err) {
  const { res, callback, body, trailers, context, opaque } = this

  removeSignal(this)

  if (callback) {
    // TODO: Does this need queueMicrotask?
    this.callback = null
    queueMicrotask(() => {
      if (this.throwOnError) {
        this.runInAsyncScope(callback, null, err, { opaque })
      }
      else {
        const { statusCode, headers } = err
        this.runInAsyncScope(callback, null, null, {
          statusCode,
          headers,
          trailers,
          opaque,
          body,
          context
        })
      }
    })
  }

  if (res) {
    this.res = null
    // Ensure all queued handlers are invoked before destroying res.
    queueMicrotask(() => {
      util.destroy(res, err)
    })
  }

  if (body) {
    this.body = null
    util.destroy(body, err)
  }
}
```

**Comment by mikaelkaron** [comment]
Another version of `onError` could be:

```javascript
onError (err) {
  const { res, callback, body, opaque } = this

  removeSignal(this)

  if (callback) {
    // TODO: Does this need queueMicrotask?
    this.callback = null
    queueMicrotask(() => {
      if (this.throwOnError) {
        this.runInAsyncScope(callback, null, err, { opaque })
      }
      else {
        this.runInAsyncScope(callback, null, null, {
         ...err,
          opaque
        })
      }
    })
  }

  if (res) {
    this.res = null
    // Ensure all queued handlers are invoked before destroying res.
    queueMicrotask(() => {
      util.destroy(res, err)
    })
  }

  if (body) {
    this.body = null
    util.destroy(body, err)
  }
}
```

**Comment by mikaelkaron** [comment]
My particular use-case is that I'm using a [`RetryHandler`](https://undici.nodejs.org/#/docs/api/RetryHandler) but I don't want to have my request rejected if the [default `onRetry`](https://github.com/nodejs/undici/blob/v6.10.0/lib/handler/retry-handler.js#L104) fails.

**Reply by metcoder95** [reply]
Can you elaborate what do you mean with `fails`?

**Reply by mikaelkaron** [reply]
In my case I’m writing a digest auth interceptor similar to [`undici-oidc-interceptor`](https://github.com/platformatic/undici-oidc-interceptor) where I need the `WWW-Authenticate` header from a `401` request to do some calculations - for this I’m using a `RetryHandler` but if authentication fails I don’t want it to throw (reject) but just pass the `401` as if it was not intercepted.

The current implementation of `RetryHandler` will call `handler.onError` when all retries have been exhausted which will throw (reject). 

This could be avoided by implementing my own `onError` but it would largely just re-implement the original code except for how the `callback` arguments are passed.

**Reply by metcoder95** [reply]
Got it, see what you mean, using a `request` won't work most likely, maybe you'll need to create a Dispatcher that wraps that logic, you can take advantage of the new `Dispatcher#compose` method to do that

**Reply by mikaelkaron** [reply]
That’s what I’m doing ;) @mcollina pointed me in that direction earlier. I’ll post a link to my WIP so you can see what’s up.

**Reply by mikaelkaron** [reply]
My WIP is available at https://github.com/mikaelkaron/undici-digest-interceptor - as you can see in the test added in https://github.com/mikaelkaron/undici-digest-interceptor/commit/37fb71f5a2ae9a28bc11d04274db9040668c76e8 it will reject (because retries are exhausted) rather than just passing the `401` back.

**Reply by mikaelkaron** [reply]
Another option would be to focus the change to be in `RetryHandler` only, but I can't figure out how to do that in a good way at the moment (or at least in a way that I don't have to rewrite a bunch of tests).

**Reply by metcoder95** [reply]
The scope of the `RetryHandler` seems good enough, the wrapper you did in the form of an interceptor is great and the way I'd  follow

**Reply by mikaelkaron** [reply]
Yeah. I guess now the problem is more along “how do I make a `RetryHandler` not error (reject) when all retries are exhausted”

**Reply by metcoder95** [reply]
Maybe some sort of extra callback that can be added when creating the handler that receives the full response in exchange, but that will remove the magic from the handler

**Reply by mikaelkaron** [reply]
I've been looking at the code and I can't see any obvious ways of telling it to "stop retry and just pass data" in a simple way (actually, the whole retry continues until `callback` is passed with an error).

A workaround _could_ be to call `retryHandler.handler.onHeaders` from inside the `retry` method (my own code), but it would be nice if `RetryHandler` had something that would allow me to just "continue"

perhaps the signature of `callback` (`onRetry`) would be `onRetry(err, data)` and if `!err` the just pass on `data`?

**Reply by metcoder95** [reply]
That's doable; what puts me over the fence is the goal of the `RetryHandler` itself and what we are trying to do.
Overall the `RetryHandler` was designed to handle automatic retries on errors (either network or responses) but adding the possibility to to bypass that automatic retry and forward an error as a simple response seems counterintuitive, but let's give it a try, I might be wrong here.

**Reply by mikaelkaron** [reply]
I see it more like “how do you want to handle failure” where it’s ok to reject with a detailed error, but it would be just as ok to pass the error downstream (or even modify it).

same function, more options (as both of them stop retries).
